File: assignments/haskell/bank-account/bank-account_test.hs
import Test.HUnit (Assertion, (@=?), runTestTT, Test(..), Counts(..))
import System.Exit (ExitCode(..), exitWith)
import BankAccount ( BankAccount, openAccount, closeAccount
                   , getBalance, incrementBalance )
import Control.Concurrent
import Control.Monad (void, replicateM)

{-
The BankAccount module should support four calls:

openAccount
  Called at the start of each test. Returns a BankAccount.

closeAccount account
  Called at the end of each test.

getBalance account
  Get the balance of the bank account.

updateBalance account amount
  Increment the balance of the bank account by the given amount.
  The amount may be negative for a withdrawal.

The initial balance of the bank account should be 0.
-}

exitProperly :: IO Counts -> IO ()
exitProperly m = do
  counts <- m
  exitWith $ if failures counts /= 0 || errors counts /= 0 then ExitFailure 1 else ExitSuccess

testCase :: String -> Assertion -> Test
testCase label assertion = TestLabel label (TestCase assertion)

main :: IO ()
main = exitProperly $ runTestTT $ TestList
  [ withBank "initial balance is 0" $
    checkReturn (Just 0) . getBalance
  , withBank "incrementing and checking balance" $ \acct -> do
    checkReturn (Just 0) $ getBalance acct
    checkReturn (Just 10) $ incrementBalance acct 10
    checkReturn (Just 10) $ getBalance acct
  , withBank "incrementing balance from other processes then checking it\
             \from test process" $ \acct -> do
    replicateM 20 (incrementProc acct) >>= mapM_ (void . takeMVar)
    checkReturn (Just 20) (getBalance acct)
  ]

incrementProc :: BankAccount -> IO (MVar ())
incrementProc acct = do
  v <- newEmptyMVar
  void . forkIO $ do
    void (incrementBalance acct 1)
    putMVar v ()
  return v

checkReturn :: (Show a, Eq a) => a -> IO a -> IO ()
checkReturn expect test = do
  v <- test
  expect @=? v

withBank :: String -> (BankAccount -> Assertion) -> Test
withBank label f = testCase label $ do
  acct <- openAccount
  f acct
  closeAccount acct

File: assignments/clojure/bank-account/bank_account_test.clj
(ns bank.test (:use clojure.test))
(load-file "bank.clj")

;; The BankAccount module should support four calls:
;;
;; open-account
;; Called at the start of each test. Returns a BankAccount.
;;
;; close-account account
;; Called at the end of each test.
;;
;; balance account
;; Get the balance of the bank account.
;;
;; update-balance account amount
;; Increment the balance of the bank account by the given amount.
;; The amount may be negative for a withdrawal.
;;
;; The initial balance of the bank account should be 0.

(defn with-bank [label f]
  (let [acct (bank/open-account)
        result (f acct label)]
    (bank/close-account acct)
    result))

(deftest checking-basic-balance
  (with-bank "initial balance is 0" #(is (= 0 (bank/get-balance %1)) %2))
  (with-bank "incrementing and checking balance"
    #(do (is (= 0 (bank/get-balance %1)) %2)
         (is (= 10 (bank/update-balance %1 10)) %2)
         (is (= 10 (bank/get-balance %1)) %2))))

(deftest can-increment-and-decrement
  (with-bank "increment and decrement balance"
    #(do (is (= 0 (bank/get-balance %1)) %2)
         (is (= 10 (bank/update-balance %1 10)) %2)
         (is (= 0 (bank/update-balance %1 -10)) %2))))

(deftest check-concurrent-access
  (with-bank "handles concurrent execution"
    #(do
      (pcalls (bank/update-balance %1 10)
              (bank/update-balance %1 10)
              (bank/update-balance %1 10))
      (is (= 30 (bank/get-balance %1)) %2))))

(run-tests)
(shutdown-agents) ;; for the pcalls above so the test exits
